MLHR-1725 #resolve Add Support for Setters to PojoUtils

vrozov opened the pull request.

243826 commented on library/src/main/java/com/datatorrent/lib/util/PojoUtils.java:
```diff
@@ -108,24 +195,32 @@ public static String upperCaseWord(String field)
-        }
-      }
-      return fieldExpression;
+      logger.debug("Class " + clazz.getName() + " does not have field " + fieldExpression + ". Will try to find getter.", ex);
```
Please use efficient syntax for logger.

vrozov replied:
Changed logger syntax.

243826 commented on library/src/main/java/com/datatorrent/lib/util/PojoUtils.java:
```diff
@@ -26,17 +26,22 @@
-  public static final String JAVA_DOT = ".";
-  public static final String DEFAULT_POJO_NAME = "pojo";
+  private static final String JAVA_DOT = ".";
+  private static final String DEFAULT_POJO_NAME = "pojo";
```
can we call this an "object" instead?

vrozov replied:
Changed DEFAULT_POJO_NAME to object based on suggestion.

tweise commented:
Vlad, please add a test for complex expression (getter and setter).

vrozov commented:
Thomas, can you point me to a documentation, JIRA or provide an example 
for complex expressions syntax.

Currently Getters only support complex expressions in a form 
"fieldA.methodB().methodC() + object.methodD().fieldE/object.methodF()". 
IMHO, syntax is confusing and will lead to errors and bugs. For example 
the following is invalid "(fieldA.methodB().methodC() + 
object.methodD().fieldE)/object.methodF()" and will throw compile error 
as "object" is always added as prefix to the expression.

Setters only support rhs assignment in complex expressions, meaning that 
the following syntax is allowed "fieldA.methodB().methodC().fieldD" as 
it is converted to "object.fieldA.methodB().methodC().fieldD = val;"

Is it necessary to support "fieldA++"? For getter it will work, but only 
if fieldA is a public field. Is it even desirable to allow such 
expressions? Generally speaking it is not a getter or a setter.

Can you please comment on generalizing getters the same way I 
generalized setters. Should I go ahead and add them to getters?

ashwinchandrap commented:
Hey Vlad, please refer to this http://javarevisited.blogspot.com/2013/08/why-use-sl4j-over-log4j-for-logging-in.html about the slf4j logger syntax. It talks about how and why its more efficient.

vrozov commented:
MLHR-1726 #comment Removed fieldListToGetExpression method as it does not provide intended functionality. The idea was to construct ".getA().getB().getC().isD()" from {"a","b","c","d"} list that latter may be used in getter as expression. To do this properly it is necessary to analyze return type of all getters. The method needs to be reimplemented once there is a use case for such functionality.

vrozov commented:
Implemented support for complex expressions in getters and setters. Use {$} for object placeholder (getters and setters) and {#} for value placeholder in setters.

Fixed boxing/unboxing issues in CassandraOutputOperator.java and MemsqlOutputOperator.java.
